### ISSUE TITLE
AKMicrophone custom input format

### ISSUE THREAD

**Issue by intonarumori** [AudioKit/AudioKit#1783]
I've spent the weekend debugging microphone issues with AudioKit and while doing so I noticed that in the `AKMicrophone` constructor the passed `format` parameter is immediately shadowed by a local `format` variable.

https://github.com/AudioKit/AudioKit/blob/master/AudioKit/Common/Nodes/Input/AKMicrophone.swift#L42

Any thoughts on this?

**Comment by aure** [comment]
That does not look good. And its formatted differently from the rest of AudioKit, lets see who added it and add them to this issue.

**Comment by aure** [comment]
Looks like @maksutovic

**Comment by maksutovic** [comment]
My understanding why I did this was solely to make sure we had a failable inutilizaré for AKMicrophone as many of us who struggled with sample rate mismatch problems. The original AKMicrophone is not written with passing a AVFormat, the only reason I added it there was the code would not compile without giving some default value which I made nil. The whole idea for this is, if getFormatForDevice fails (because of samplerate mismatch), have a way to gracefully alert the user instead of hardcrashing.

Let me know if you need anymore clarity -- Only intent was to take existing init() and make a failable one, which is what I did.

**Comment by intonarumori** [comment]
Thanks for the fast reply.

I was also struggling with format mismatch errors with the microphone and trying to debug the issue with the limited experience I have with AudioKit and AVAudioEngine.

My takeaways are:
- it's best to wire up `engine.inputNode` before starting the engine, every attempt to connect it later resulted in the format mismatch crash
- if I wire up `engine.inputNode` with a mixer before the engine is started, I was able to connect to that mixer later with another node and tap into that to get data
- Apple sample code uses `engine.inputNode.inputFormat(forBus: 0)` to get the format, but in my experience it always results in the sample rate set to `0hz` for that format (might be the reason why in `getFormatForDevice` it is overwritten with the device sample rate)

I would be very interested in hearing your experience with adding nodes to a running engine, especially if it involves connecting to the input node.
Based on the WWDC videos and documentation it should be possible to attach/detach nodes while running, but I couldn't get it working without experiencing the crash.

I understand this might be out of scope for this particular issue, but I would be happy to learn about your experience and best practices to handling input.

**Comment by intonarumori** [comment]
On further debugging I think I found solution I was looking for:

If I access the `engine.inputNode` before starting the engine everything starts working fine, no prewiring is needed.

My guess is that this forces the engine to create the `inputNode` singleton so it's probably initialised properly. If you would do this after the engine is already running the crash will likely happen.

I created a gist for a simple recorder for future reference here:
https://gist.github.com/intonarumori/8e7b56597bd5d451fd098b11c062b01e

**Comment by maksutovic** [comment]
Yes, it is important to wire `AKMicrophone` prior to starting audio kit.

**Comment by intonarumori** [comment]
@maksutovic To be precise, you don't have to wire it up, you have to get it initialised, and even that requirement is probably only due to an Apple bug.

I think it's unfortunate that every example of AudioKit code is wiring everything up before starting the engine, thus giving the false impression that this is the only way it will work.
Apple emphasised in their WWDC talk that nodes can be attached/detached/connected while the engine is running and that's what I was looking for.

**Comment by aure** [comment]
AudioKit is open source if you want to help make connections/disconnections painless!

**Comment by intonarumori** [comment]
@aure I definitely would consider contributing in the future, but for now I'm just getting into the details with both AVAudioEngine and AudioKit.

These are more like my first impressions, I'm pretty sure there are good reasons you guys decided to "hide" the details of making connections using the engine instance itself and rather advocate using connect methods of the nodes or passing inputs via constructors.

I was able to attach/detach a custom microphone node without issues, but I don't fully understand yet why it is not working with the provided one. I'll keep you updated about my findings.

**Comment by aure** [comment]
I changed it so that it only uses that format if no other format was specified.  I also added a _ = engine.inputNode like @intonarumori's solution. HTH
